lib/checkmemoryleak.cpp

```diff
@@ -546,10 +546,26 @@ void CheckMemoryLeakInFunction::checkReallocUsage()
                     continue;
 
                 // Check that another copy of the pointer wasn't saved earlier in the function
-                if (Token::findmatch(scope->bodyStart, "%name% = %varid% ;", tok, tok->varId()) ||
-                    Token::findmatch(scope->bodyStart, "[{};] %varid% = *| %name% .| %name%| [;=]", tok, tok->varId()))
+                if (Token::findmatch(scope->bodyStart, "%name% = %varid% ;", tok, tok->varId()))
                     continue;
 
+                bool skip = false;
+                const Token* tok1 = scope->bodyStart;
+                while ((tok1 = Token::findmatch(tok1, "[{};] %varid% = *| %name% .| %name%| [;=]", tok, tok->varId()))) {
+                    if (!Token::Match(tok1->tokAt(3), "NULL|nullptr")) {
```

Comment:
as you say this is a little cludgy.. first of all the loop is stupid. and the condition NULL|nullptr is hardcoded. and I fear we have some similar code here and there. I have the feeling that instead of the loop it would be useful with a utility function that extracts possible modifications in different paths for a local variable.
not sure if that needs to be added right now in this PR though.. we could merge this and then perform refactorings later.

Reply:
I needed the loop because I didn't want to warn for the following case
```c
void f(char *b) {
    void *a = NULL;
    a = b;
    a = realloc(a, 20);
} 
```
Here, the current `Token::findmatch()` will find the first assignment and not the second so I needed to check all matches (hence the loop), but what I really wanted was to match on variables but not `NULL` or `nullptr`. It seems it is sufficient to change to match on `%var%` instead of `%name%` (at least to get the tests passing), let me run a few more tests...

Reply:
another solution better in the long term could be to rewrite this check to use `FwdAnalysis`

Reply:
>  It seems it is sufficient to change to match on %var% instead of %name% (at least to get the tests passing), let me run a few more tests

that sounds interesting.

Reply:
I pushed that instead. I ran it on 138 packages with `test-my-pr.py` and it did not cause any new false positives (but 19 more TPs (of which 10 in the same package) in 6 packages)).